# Day 2: Rock Paper Scissors

The Elves begin to set up camp on the beach. To decide whose tent gets to be closest to the snack storage, a giant Rock Paper Scissors tournament is already in progress.

Rock Paper Scissors is a game between two players. Each game contains many rounds; in each round, the players each simultaneously choose one of Rock, Paper, or Scissors using a hand shape. Then, a winner for that round is selected: Rock defeats Scissors, Scissors defeats Paper, and Paper defeats Rock. If both players choose the same shape, the round instead ends in a draw.

Appreciative of your help yesterday, one Elf gives you an encrypted strategy guide (your puzzle input) that they say will be sure to help you win. "The first column is what your opponent is going to play: A for Rock, B for Paper, and C for Scissors. The second column--" Suddenly, the Elf is called away to help with someone's tent.

The second column, you reason, must be what you should play in response: X for Rock, Y for Paper, and Z for Scissors. Winning every time would be suspicious, so the responses must have been carefully chosen.

The winner of the whole tournament is the player with the highest score. Your total score is the sum of your scores for each round. The score for a single round is the score for the shape you selected (1 for Rock, 2 for Paper, and 3 for Scissors) plus the score for the outcome of the round (0 if you lost, 3 if the round was a draw, and 6 if you won).

Since you can't be sure if the Elf is trying to help you or trick you, you should calculate the score you would get if you were to follow the strategy guide.

For example, suppose you were given the following strategy guide:

```
A Y
B X
C Z
```

This strategy guide predicts and recommends the following:

- In the first round, your opponent will choose Rock (A), and you should choose Paper (Y). This ends in a win for you with a score of 8 (2 because you chose Paper + 6 because you won).
- In the second round, your opponent will choose Paper (B), and you should choose Rock (X). This ends in a loss for you with a score of 1 (1 + 0).
- The third round is a draw with both players choosing Scissors, giving you a score of 3 + 3 = 6.

In this example, if you were to follow the strategy guide, you would get a total score of 15 (8 + 1 + 6).

What would your total score be if everything goes exactly according to your strategy guide?

In [60]:
#@title Set up some constants for readability

class OpponentMove:
    ROCK = 'A'
    PAPER = 'B'
    SCISSORS = 'C'

class PlayerMove:
    ROCK = 'X'
    PAPER = 'Y'
    SCISSORS = 'Z'

class MoveScore:
    ROCK = 1
    PAPER = 2
    SCISSORS = 3

class ResultScore:
    WIN = 6
    LOSE = 0
    DRAW = 3

In [61]:
#@title Function to calculate score for the round

"""
The score for a single round is the score for the shape you selected 
(1 for Rock, 2 for Paper, and 3 for Scissors) plus the score for the 
outcome of the round (0 if you lost, 3 if the round was a draw, and 
6 if you won)
"""
def calculate_score(opponent: str, player: str) -> int:

    # !note! on python match/case:
    # A variable name in a case clause is treated as a name capture pattern (https://www.python.org/dev/peps/pep-0634/#capture-patterns).
    # It always matches and tries to make an assignment to the variable name. This is not what we need so we put the enums in a class and 
    # as per https://stackoverflow.com/questions/67525257/capture-makes-remaining-patterns-unreachable.
    match player: 
        case PlayerMove.ROCK: 
            match opponent:
                case OpponentMove.ROCK:
                    return (MoveScore.ROCK + ResultScore.DRAW)
                case OpponentMove.PAPER:
                    return (MoveScore.ROCK + ResultScore.LOSE)
                case OpponentMove.SCISSORS:
                    return (MoveScore.ROCK + ResultScore.WIN)
        case PlayerMove.PAPER: 
            match opponent:
                case OpponentMove.ROCK:
                    return (MoveScore.PAPER + ResultScore.WIN)
                case OpponentMove.PAPER:
                    return (MoveScore.PAPER + ResultScore.DRAW)
                case OpponentMove.SCISSORS:
                    return (MoveScore.PAPER + ResultScore.LOSE)
        case PlayerMove.SCISSORS: 
            match opponent:
                case OpponentMove.ROCK:
                    return (MoveScore.SCISSORS + ResultScore.LOSE)
                case OpponentMove.PAPER:
                    return (MoveScore.SCISSORS + ResultScore.WIN)
                case OpponentMove.SCISSORS:
                    return (MoveScore.SCISSORS + ResultScore.DRAW)

In [62]:
#@title Test calculate_score function
import unittest

class TestCalculateScore(unittest.TestCase):
    
    def test_rock_rock(self):
        self.assertEqual(calculate_score(OpponentMove.ROCK, PlayerMove.ROCK), 4)
    
    def test_paper_paper(self):
        self.assertEqual(calculate_score(OpponentMove.PAPER, PlayerMove.PAPER), 5)

    def test_scissors_scissors(self):
        self.assertEqual(calculate_score(OpponentMove.SCISSORS, PlayerMove.SCISSORS), 6)

    def test_rock_paper(self):
        self.assertEqual(calculate_score(OpponentMove.PAPER, PlayerMove.ROCK), 1)

    def test_rock_scissors(self):
        self.assertEqual(calculate_score(OpponentMove.SCISSORS, PlayerMove.ROCK), 7)

    def test_paper_rock(self):
        self.assertEqual(calculate_score(OpponentMove.ROCK, PlayerMove.PAPER), 8)

    def test_paper_scissors(self):
        self.assertEqual(calculate_score(OpponentMove.SCISSORS, PlayerMove.PAPER), 2)    

    def test_scissors_paper(self):
        self.assertEqual(calculate_score(OpponentMove.PAPER, PlayerMove.SCISSORS), 9)

    def test_scissors_rock(self):
        self.assertEqual(calculate_score(OpponentMove.ROCK, PlayerMove.SCISSORS), 3)


unittest.main(argv=[''], verbosity=2, exit=False)

test_paper_paper (__main__.TestCalculateScore.test_paper_paper) ... ok
test_paper_rock (__main__.TestCalculateScore.test_paper_rock) ... ok
test_paper_scissors (__main__.TestCalculateScore.test_paper_scissors) ... ok
test_rock_paper (__main__.TestCalculateScore.test_rock_paper) ... ok
test_rock_rock (__main__.TestCalculateScore.test_rock_rock) ... ok
test_rock_scissors (__main__.TestCalculateScore.test_rock_scissors) ... ok
test_scissors_paper (__main__.TestCalculateScore.test_scissors_paper) ... ok
test_scissors_rock (__main__.TestCalculateScore.test_scissors_rock) ... ok
test_scissors_scissors (__main__.TestCalculateScore.test_scissors_scissors) ... ok

----------------------------------------------------------------------
Ran 9 tests in 0.012s

OK


In [63]:
#@title Imports
import pandas as pd

In [64]:
#@title import the data
df = pd.read_csv('input.txt', sep=' ', header=None)

In [65]:
#@title Clean and convert the data

# Add Column Names
df.columns = ['Opponent Letter', 'Player Letter']

# add score
df['Score'] = df.apply(lambda x: calculate_score(x['Opponent Letter'], x['Player Letter']), axis=1)


In [66]:
df

Opponent Letter Player Letter  Score
0                  C             X      7
1                  C             Y      2
2                  C             X      7
3                  B             X      1
4                  B             Z      9
...              ...           ...    ...
2495               B             Z      9
2496               C             X      7
2497               C             Y      2
2498               A             Z      3
2499               A             Z      3

[2500 rows x 3 columns]

In [67]:
#@title Get the answer!

df["Score"].sum()

13484

Your puzzle answer was 13484.

The first half of this puzzle is complete! It provides one gold star: *

## Part Two

The Elf finishes helping with the tent and sneaks back over to you. "Anyway, the second column says how the round needs to end: X means you need to lose, Y means you need to end the round in a draw, and Z means you need to win. Good luck!"

```
A Y
B X
C Z
```

The total score is still calculated in the same way, but now you need to figure out what shape to choose so the round ends as indicated. The example above now goes like this:

- In the first round, your opponent will choose Rock (A), and you need the round to end in a draw (Y), so you also choose Rock. This gives you a score of 1 + 3 = 4.
- In the second round, your opponent will choose Paper (B), and you choose Rock so you lose (X) with a score of 1 + 0 = 1.
- In the third round, you will defeat your opponent's Scissors with Rock for a score of 1 + 6 = 7.

Now that you're correctly decrypting the ultra top secret strategy guide, you would get a total score of 12.

Following the Elf's instructions for the second column, what would your total score be if everything goes exactly according to your strategy guide?